File: packages/cli/hieudoanm/swift/Sources/hieudoanm/Commands/Chess/Chess.swift
import Foundation
import ArgumentParser

func resolveCountry(_ url: String) -> (code: String, name: String) {
    if url.isEmpty { return ("-", "-") }
    let trimmed = url.hasSuffix("/") ? String(url.dropLast()) : url
    let parts = trimmed.split(separator: "/")
    guard let last = parts.last else { return ("-", "-") }
    let code = last.uppercased()
    if let name = COUNTRIES[code] { return (code, name) }
    return (code, "-")
}

struct ChessCommand: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "chess",
        abstract: "Chess tools",
        subcommands: [
            ChessElo.self, ChessFENEval.self, ChessFENSVG.self,
            Chess960.self, ChessSetup.self, ChessComPlayer.self,
            ChessComLeaderboards.self, ChessComTitled.self,
        ]
    )
    mutating func run() {}
}

struct ChessElo: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "elo", abstract: "Calculate ELO rating change")
    @Argument(help: "Player A rating") var ratingA: Int
    @Argument(help: "Player B rating") var ratingB: Int
    @Argument(help: "Result (1=A wins, 0=B wins, 0.5=draw)") var result: Double
    mutating func run() {
        let expectedA = 1.0 / (1.0 + pow(10.0, Double(ratingB - ratingA) / 400.0))
        let expectedB = 1.0 - expectedA
        let newA = Double(ratingA) + 32.0 * (result - expectedA)
        let newB = Double(ratingB) + 32.0 * ((1.0 - result) - expectedB)
        print(String(format: "Expected: A=%.3f, B=%.3f", expectedA, expectedB))
        print("New ratings: A=\(Int(newA)) (\(Int(newA - Double(ratingA)))), B=\(Int(newB)) (\(Int(newB - Double(ratingB))))")
    }
}

struct ChessFENEval: AsyncParsableCommand {
    static let configuration = CommandConfiguration(commandName: "fen-eval", abstract: "Evaluate FEN position via Lichess cloud")
    @Argument(help: "FEN string") var fen: String
    mutating func run() async throws {
        let encoded = fen.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? fen
        let url = "https://explorer.lichess.ovh/masters?fen=\(encoded)"
        let response = try await Requests.fetch(url)
        let json = try JSONSerialization.jsonObject(with: response.data) as? [String: Any]
        if let moves = json?["moves"] as? [[String: Any]] {
            print("Top moves (Lichess Masters):")
            for move in moves.prefix(5) {
                let uci = move["uci"] as? String ?? ""
                let san = move["san"] as? String ?? ""
                let white = move["white"] as? Int ?? 0
                let black = move["black"] as? Int ?? 0
                let draws = move["draws"] as? Int ?? 0
                let total = white + black + draws
                let rate = total > 0 ? Double(white) / Double(total) * 100 : 0
                print(String(format: "  %@ (%@): W%d L%d D%d win rate %.1f%%", san, uci, white, black, draws, rate))
            }
        } else {
            print("No evaluation data available for this position")
        }
    }
}

struct ChessFENSVG: AsyncParsableCommand {
    static let configuration = CommandConfiguration(commandName: "fen-svg", abstract: "Get FEN board SVG URL")
    @Argument(help: "FEN string") var fen: String
    mutating func run() async throws {
        let encoded = fen.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? fen
        let url = "https://chess.com/board.png?fen=\(encoded)"
        print("Board image URL: \(url)")
        let response = try await Requests.fetch(url)
        let filename = "chessboard.png"
        try response.data.write(to: URL(fileURLWithPath: filename))
        print("Saved to \(filename)")
    }
}

struct Chess960: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "random", abstract: "Generate Chess960 position")
    mutating func run() {
        var pos = [String](repeating: "", count: 8)
        var avail = Array(0..<8)
        let light = avail.filter { $0 % 2 == 0 }; let dark = avail.filter { $0 % 2 == 1 }
        pos[light.randomElement()!] = "B"; pos[dark.randomElement()!] = "B"
        avail = avail.filter { pos[$0].isEmpty }
        pos[avail.randomElement()!] = "Q"; avail = avail.filter { pos[$0].isEmpty }
        for _ in 0..<2 { let i = avail.randomElement()!; pos[i] = "N"; avail = avail.filter { pos[$0].isEmpty } }
        let r = avail.sorted()
        pos[r[0]] = "R"; pos[r[1]] = "K"; pos[r[2]] = "R"
        print(pos.joined(separator: " "))
    }
}

struct ChessSetup: ParsableCommand {
    static let configuration = CommandConfiguration(commandName: "setup", abstract: "Piece placement puzzle")
    @Option(name: .long, help: "Number of pieces") var count: Int = 3
    mutating func run() {
        let files = "abcdefgh"
        print("Place these pieces on the board:")
        for _ in 0..<count {
            let piece = ["♔", "♕", "♖", "♗", "♘", "♙"].randomElement()!
            let file = files.randomElement()!
            let rank = Int.random(in: 1...8)
            print("  \(piece) \(file)\(rank)")
        }
    }
}

struct Player: Decodable {
    let rank: Int
    let username: String
    let name: String?
    let score: Int
    let country: String
    let title: String?
    let winCount: Int
    let drawCount: Int
    let lossCount: Int
}

struct LeaderboardsResponse: Decodable {
    let liveBullet: [Player]
    let liveBlitz: [Player]
    let liveRapid: [Player]
    let liveBlitz960: [Player]
}

struct ChessComLeaderboards: AsyncParsableCommand {
    static let configuration = CommandConfiguration(commandName: "com-leaderboards", abstract: "Fetch Chess.com leaderboards")
    @Option(name: .long, help: "Number of top players to display") var top: Int = 5
    @Option(name: .long, help: "Filter by country code (e.g. US, RU)") var country: String = ""

    mutating func run() async throws {
        let response = try await Requests.fetch("https://api.chess.com/pub/leaderboards")
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        let data = try decoder.decode(LeaderboardsResponse.self, from: response.data)

        let countryFilter = country
        let filterPlayers: ([Player]) -> [Player] = { players in
            guard !countryFilter.isEmpty else { return players }
            return players.filter { p in
                let (code, _) = resolveCountry(p.country)
                return code == countryFilter.uppercased()
            }
        }

        printTop("Live Bullet", filterPlayers(data.liveBullet), top)
        printTop("Live Blitz", filterPlayers(data.liveBlitz), top)
        printTop("Live Rapid", filterPlayers(data.liveRapid), top)
        printTop("Live Blitz 960", filterPlayers(data.liveBlitz960), top)
    }
}

private func printTop(_ title: String, _ players: [Player], _ limit: Int) {
    guard !players.isEmpty else { return }
    print("\n\(title)\n")
    let count = min(limit, players.count)
    for i in 0..<count {
        let p = players[i]
        let name = p.name ?? "-"
        let (_, countryName) = resolveCountry(p.country)
        let wdl = "\(p.winCount) / \(p.drawCount) / \(p.lossCount)"
        print("\(String(format: "%3d", p.rank)). \(name) (\(p.username.lowercased())) - \(countryName) - \(p.score) - \(wdl)")
    }
}

struct ChessComPlayer: AsyncParsableCommand {
    static let configuration = CommandConfiguration(commandName: "com-player", abstract: "Fetch Chess.com player stats")
    @Argument(help: "Username") var username: String
    mutating func run() async throws {
        let username = username.lowercased()
        let profileURL = "https://api.chess.com/pub/player/\(username)"
        let statsURL = "https://api.chess.com/pub/player/\(username)/stats"

        async let profileResp = try Requests.fetch(profileURL)
        async let statsResp = try Requests.fetch(statsURL)
        let (profileData, statsData) = try await (profileResp, statsResp)

        let profile = try JSONSerialization.jsonObject(with: profileData.data) as? [String: Any] ?? [:]
        let stats = try JSONSerialization.jsonObject(with: statsData.data) as? [String: Any] ?? [:]

        print()
        print("Player: \(username.uppercased())")
        print("---------------------------------------------------------------")
        if let name = profile["name"] as? String, !name.isEmpty { print("Name      : \(name)") }
        if let title = profile["title"] as? String, !title.isEmpty { print("Title     : \(title)") }
        if let countryURL = profile["country"] as? String {
            let (_, countryName) = resolveCountry(countryURL)
            print("Country   : \(countryName)")
        }
        if let fide = profile["fide"] as? Int, fide > 0 { print("FIDE      : \(fide)") }
        if let followers = profile["followers"] as? Int { print("Followers : \(followers)") }

        print("\nRatings\n")
        printRatingsHeader()
        printRating("Bullet", stats["chess_bullet"] as? [String: Any])
        printRating("Blitz", stats["chess_blitz"] as? [String: Any])
        printRating("Rapid", stats["chess_rapid"] as? [String: Any])
    }
}

private func printRatingsHeader() {
    print("| Mode    |    Best |    Last |     Win |    Draw |    Loss |")
    print("|---------|--------|--------|--------|--------|--------|")
}

private func printRating(_ label: String, _ stats: [String: Any]?) {
    let best = stats?["best"] as? [String: Any] ?? [:]
    let last = stats?["last"] as? [String: Any] ?? [:]
    let record = stats?["record"] as? [String: Any] ?? [:]
    let bestR = best["rating"] as? Int ?? 0
    let lastR = last["rating"] as? Int ?? 0
    let wins = record["wins"] as? Int ?? 0
    let draws = record["draws"] as? Int ?? 0
    let losses = record["losses"] as? Int ?? 0
    print(String(format: "| %-7s | %6d | %6d | %6d | %6d | %6d |", label, bestR, lastR, wins, draws, losses))
}

struct ChessComTitled: AsyncParsableCommand {
    static let configuration = CommandConfiguration(commandName: "com-titled", abstract: "Fetch Chess.com titled players")
    @Argument(help: "Title (GM, IM, FM, etc.)") var title: String
    mutating func run() async throws {
        let response = try await Requests.fetch("https://api.chess.com/pub/titled/\(title.uppercased())")
        let json = try JSONSerialization.jsonObject(with: response.data) as? [String: Any]
        let players = json?["players"] as? [String] ?? []
        print("\(title.uppercased()) players: \(players.count)")
        for name in players.prefix(20) { print("  \(name)") }
        if players.count > 20 { print("  ... and \(players.count - 20) more") }
    }
}

File: packages/cli/jack/cli.kt/src/main/kotlin/io/github/hieudoanm/cli/commands/Chess.kt
package io.github.hieudoanm.cli.commands

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.subcommands
import com.github.ajalt.clikt.parameters.arguments.argument
import com.github.ajalt.clikt.parameters.arguments.default
import com.github.ajalt.clikt.parameters.options.option
import com.github.ajalt.clikt.parameters.options.flag
import com.github.ajalt.clikt.parameters.options.default
import com.github.ajalt.clikt.parameters.options.required
import com.github.ajalt.clikt.parameters.types.int
import com.github.ajalt.clikt.parameters.types.double
import io.github.hieudoanm.cli.services.Requests
import com.google.gson.Gson
import java.io.File
import kotlin.math.roundToInt
import kotlin.random.Random
import kotlin.math.*

class ChessCommand : CliktCommand(name = "chess", help = "Chess tools and utilities") {
    init {
        subcommands(ChessCom(), ChessElo(), ChessFenEval(), ChessFenSvg(), ChessPgn(), ChessPlay(), ChessRandom(), ChessSetup())
    }
    override fun run() = Unit
}

class ChessCom : CliktCommand(name = "com", help = "Chess.com integration") {
    init { subcommands(ChessComPlayerCmd(), ChessComLeaderboards(), ChessComTitled()) }
    override fun run() = Unit
}

class ChessComPlayerCmd : CliktCommand(name = "player", help = "Show Chess.com player profile and stats") {
    private val username by option("--username", "-u", help = "Chess.com username").required()
    override fun run() {
        val u = username.lowercase()
        val profileResult = Requests.get("https://api.chess.com/pub/player/$u")
        val statsResult = Requests.get("https://api.chess.com/pub/player/$u/stats")

        profileResult.onSuccess { profileBody ->
            statsResult.onSuccess { statsBody ->
                val gson = Gson()
                val profile = gson.fromJson(profileBody, ChessComPlayerProfile::class.java)
                val stats = gson.fromJson(statsBody, ChessComPlayerStats::class.java)

                echo()
                echo("${profile.username.uppercase()}")
                echo("---------------------------------------------------------------")
                if (profile.name != null) echo("Name      : ${profile.name}")
                if (profile.title != null) echo("Title     : ${profile.title}")
                val countryCode = profile.country?.split("/")?.lastOrNull()?.uppercase() ?: "-"
                echo("Country   : $countryCode")
                if (profile.fide > 0) echo("FIDE      : ${profile.fide}")
                echo("Followers : ${profile.followers}")
                echo()
                echo("Ratings")
                echo()
                echo("%-8s | %8s | %8s | %8s | %8s | %8s".format("Mode", "Best", "Last", "Win", "Draw", "Loss"))
                echo("%-8s | %8s | %8s | %8s | %8s | %8s".format("-".repeat(8), "-".repeat(8), "-".repeat(8), "-".repeat(8), "-".repeat(8), "-".repeat(8)))
                fun printRating(label: String, r: ChessComRating?) {
                    if (r != null) {
                        echo("%-8s | %8s | %8s | %8s | %8s | %8s".format(label, r.best.rating, r.last.rating, r.record.win, r.record.draw, r.record.loss))
                    }
                }
                printRating("Bullet", stats.chessBullet)
                printRating("Blitz", stats.chessBlitz)
                printRating("Rapid", stats.chessRapid)
            }
        }
    }
}

data class ChessComPlayerProfile(val username: String, val name: String?, val title: String?, val country: String?, val fide: Int, val followers: Int)
data class ChessComRatingLast(val rating: Int, val date: Int, val rd: Int)
data class ChessComRatingBest(val rating: Int, val date: Int, val game: String?)
data class ChessComRecord(val win: Int, val draw: Int, val loss: Int)
data class ChessComRating(val last: ChessComRatingLast, val best: ChessComRatingBest, val record: ChessComRecord)
data class ChessComPlayerStats(val chessBullet: ChessComRating?, val chessBlitz: ChessComRating?, val chessRapid: ChessComRating?)

class ChessComLeaderboards : CliktCommand(name = "leaderboards", help = "Show Chess.com leaderboards") {
    private val top by option("--top", help = "Number of top players").int().default(5)
    private val countryFilter by option("--country", help = "Filter by country code").default("")
    override fun run() {
        val result = Requests.get("https://api.chess.com/pub/leaderboards")
        result.onSuccess { body ->
            val gson = Gson()
            val data = gson.fromJson(body, ChessComLeaderboardData::class.java)
            val limit = maxOf(1, top)
            fun countryName(url: String): String {
                val code = url.trimEnd('/').split("/").lastOrNull()?.uppercase() ?: "-"
                return code
            }
            fun filterPlayers(players: List<ChessComPlayer>): List<ChessComPlayer> {
                if (countryFilter.isBlank()) return players
                return players.filter { p ->
                    val code = countryName(p.country)
                    code == countryFilter.uppercase()
                }
            }
            fun printTop(title: String, players: List<ChessComPlayer>) {
                if (players.isEmpty()) return
                val filtered = filterPlayers(players)
                val shown = filtered.take(limit)
                if (shown.isEmpty()) return
                echo()
                echo(title)
                echo()
                echo("%-4s | %-32s | %-24s | %-6s".format("Rank", "Name", "Username", "Score"))
                echo("%-4s | %-32s | %-24s | %-6s".format("-".repeat(4), "-".repeat(32), "-".repeat(24), "-".repeat(6)))
                shown.forEach { p ->
                    val name = if (p.name.isNullOrBlank()) "-" else p.name
                    echo("%-4d | %-32s | %-24s | %-6d".format(p.rank, name, p.username.lowercase(), p.score))
                }
            }
            printTop("Live Bullet", data.liveBullet)
            printTop("Live Blitz", data.liveBlitz)
            printTop("Live Rapid", data.liveRapid)
            printTop("Live Blitz 960", data.liveBlitz960)
        }.onFailure { e -> echo("Error: ${e.message}") }
    }
}

data class ChessComPlayer(val rank: Int, val username: String, val name: String?, val score: Int, val country: String, val winCount: Int, val drawCount: Int, val lossCount: Int)
data class ChessComLeaderboardData(val liveBullet: List<ChessComPlayer>, val liveBlitz: List<ChessComPlayer>, val liveRapid: List<ChessComPlayer>, val liveBlitz960: List<ChessComPlayer>)

class ChessComTitled : CliktCommand(name = "titled", help = "Show Chess.com titled player counts") {
    private val titles = listOf("GM", "IM", "FM", "CM", "NM", "WGM", "WIM", "WFM", "WCM", "WNM")
    override fun run() {
        echo()
        echo("%-6s | %7s".format("Titled", "Players"))
        echo("%-6s | %7s".format("-".repeat(6), "-".repeat(7)))
        for (title in titles) {
            val result = Requests.get("https://api.chess.com/pub/titled/$title")
            result.onSuccess { body ->
                val resp = Gson().fromJson(body, ChessComTitledResponse::class.java)
                echo("%-6s | %7d".format(title, resp.players.size))
            }.onFailure { e ->
                echo("Failed to fetch $title: ${e.message}")
            }
        }
        echo()
    }
}

data class ChessComTitledResponse(val players: List<String>)

class ChessElo : CliktCommand(name = "elo", help = "Calculate new Elo rating after a game") {
    private val myRating by option("--rating", "-r", help = "Your rating").int().required()
    private val opponentRating by option("--opponent", "-o", help = "Opponent's rating").int().required()
    private val score by option("--score", "-s", help = "Result (1=win, 0.5=draw, 0=loss)").double().required()
    private val kFactor by option("--k", "-k", help = "K-factor").int().default(20)
    override fun run() {
        val expected = 1.0 / (1.0 + 10.0.pow((opponentRating - myRating) / 400.0))
        val newRating = myRating + kFactor * (score - expected)
        echo("Your new rating: ${"%.0f".format(newRating)}")
    }
}

class ChessFenEval : CliktCommand(name = "fen-eval", help = "Evaluate a FEN position using Lichess cloud eval") {
    private val fen by option("--fen", "-f", help = "FEN string").required()
    private val multipv by option("--multipv", help = "Number of principal variations").int().default(3)
    override fun run() {
        val url = "https://lichess.org/api/cloud-eval?fen=${java.net.URLEncoder.encode(fen, "UTF-8")}&multiPv=$multipv&variant=standard"
        val result = Requests.get(url)
        result.onSuccess { body ->
            val gson = Gson()
            val eval = gson.fromJson(body, CloudEvalResponse::class.java)
            echo()
            echo("lichess.org Cloud Evaluation")
            echo("------------------------------------------------")
            echo("Depth : ${eval.depth}")
            echo("Nodes : ${eval.knodes}")
            echo()
            eval.pvs.forEachIndexed { i, pv ->
                val cp = if (pv.cp != null) "${"%.2f".format(pv.cp / 100.0)}" else if (pv.mate != null) "#${pv.mate}" else "?"
                echo("#${i + 1}  $cp  ${pv.moves}")
            }
        }.onFailure { e -> echo("Failed to fetch evaluation: ${e.message}") }
    }
}

data class CloudEvalPV(val moves: String, val cp: Int?, val mate: Int?)
data class CloudEvalResponse(val fen: String, val depth: Int, val knodes: Int, val pvs: List<CloudEvalPV>)

class ChessFenSvg : CliktCommand(name = "fen-svg", help = "Render a FEN position as an SVG board image") {
    private val fen by option("--fen", "-f", help = "FEN string").required()
    private val out by option("--out", "-o", help = "Output SVG file").default("board.svg")
    override fun run() {
        val svg = renderSvgFromFen(fen)
        File(out).writeText(svg)
        echo("SVG saved to $out")
    }
}

private fun renderSvgFromFen(fen: String): String {
    val pieceMap = mapOf(
        'K' to "\u2654", 'Q' to "\u2655", 'R' to "\u2656", 'B' to "\u2657", 'N' to "\u2658", 'P' to "\u2659",
        'k' to "\u265A", 'q' to "\u265B", 'r' to "\u265C", 'b' to "\u265D", 'n' to "\u265E", 'p' to "\u265F"
    )
    val boardPart = fen.split(" ").first()
    val ranks = boardPart.split("/")
    val squareSize = 60
    val boardSize = squareSize * 8
    val sb = StringBuilder()
    sb.appendLine("<svg xmlns=\"http://www.w3.org/2000/svg\" width=\"$boardSize\" height=\"$boardSize\">")
    for (rank in 0 until 8) {
        for (file in 0 until 8) {
            val x = file * squareSize
            val y = (7 - rank) * squareSize
            val color = if ((rank + file) % 2 == 1) "#f0d9b5" else "#b58863"
            sb.appendLine("<rect x=\"$x\" y=\"$y\" width=\"$squareSize\" height=\"$squareSize\" fill=\"$color\"/>")
        }
    }
    for (rank in 0 until 8) {
        var file = 0
        for (ch in ranks.getOrElse(rank) { "" }) {
            if (ch.isDigit()) {
                file += ch - '0'
            } else {
                val x = file * squareSize
                val y = (7 - rank) * squareSize
                val p = pieceMap[ch] ?: ch.toString()
                sb.appendLine("<text x=\"${x + squareSize / 2}\" y=\"${y + squareSize * 0.75}\" text-anchor=\"middle\" font-size=\"${squareSize * 0.75}\">$p</text>")
                file++
            }
        }
    }
    sb.appendLine("</svg>")
    return sb.toString()
}

class ChessPgn : CliktCommand(name = "pgn", help = "PGN chess game analysis tools") {
    init { subcommands(ChessPgnFen(), ChessPgnUci()) }
    override fun run() = Unit
}

class ChessPgnFen : CliktCommand(name = "fen", help = "Convert PGN to FEN per move") {
    private val pgnFile by option("--pgn-file", help = "Path to PGN file").default("")
    private val pgn by option("--pgn", help = "Raw PGN string").default("")
    override fun run() {
        val pgnStr = when {
            pgnFile.isNotBlank() -> File(pgnFile).readText()
            pgn.isNotBlank() -> pgn
            else -> throw Exception("provide --pgn-file or --pgn")
        }
        val moves = pgnStr.replace(Regex("""\d+\."""), "").replace(Regex("""[\[\].\n\r]+"""), " ").trim()
            .split(Regex("\\s+")).filter { it.isNotBlank() && !it.contains("?") && !it.contains("!") }
        echo("%-4s | %-6s | %-72s".format("Move", "SAN", "FEN"))
        echo("%-4s | %-6s | %-72s".format("-".repeat(4), "-".repeat(6), "-".repeat(72)))
        moves.forEachIndexed { i, m ->
            val moveNum = i / 2 + 1
            val side = if (i % 2 == 0) "White" else "Black"
            echo("%-4d | %-6s | %-72s".format(moveNum, m, "-"))
        }
    }
}

class ChessPgnUci : CliktCommand(name = "uci", help = "Convert PGN moves to UCI notation") {
    private val pgnFile by option("--pgn-file", help = "Path to PGN file").default("")
    private val pgn by option("--pgn", help = "Raw PGN string").default("")
    override fun run() {
        val pgnStr = when {
            pgnFile.isNotBlank() -> File(pgnFile).readText()
            pgn.isNotBlank() -> pgn
            else -> throw Exception("provide --pgn-file or --pgn")
        }
        val moves = pgnStr.replace(Regex("""\d+\."""), "").replace(Regex("""[\[\].\n\r]+"""), " ").trim()
            .split(Regex("\\s+")).filter { it.isNotBlank() && !it.contains("?") && !it.contains("!") }
        echo(moves.joinToString(" "))
    }
}

class ChessPlay : CliktCommand(name = "play", help = "Play chess interactively in the terminal") {
    private val blind by option("--blind", help = "Hide the board").flag()
    override fun run() {
        val pieces = mapOf(
            'K' to "K", 'Q' to "Q", 'R' to "R", 'B' to "B", 'N' to "N", 'P' to "P",
            'k' to "k", 'q' to "q", 'r' to "r", 'b' to "b", 'n' to "n", 'p' to "p"
        )
        var board = Array(8) { Array(8) { "" } }
        val startFen = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR"
        val ranks = startFen.split("/")
        for (r in 0 until 8) {
            var f = 0
            for (ch in ranks[r]) {
                if (ch.isDigit()) { f += ch - '0' }
                else { board[r][f] = pieces[ch] ?: ""; f++ }
            }
        }
        var turn = "White"

        echo("Chess CLI - Interactive Play")
        echo("Type moves in SAN (e.g., e4, Nf3). Type 'exit' to quit.")
        if (blind) echo("Blind mode is ON.")
        echo()

        while (true) {
            if (!blind) {
                for (r in 0 until 8) {
                    val row = board[r].joinToString(" ") { if (it.isEmpty()) "." else it }
                    echo("${8 - r} $row")
                }
                echo("  a b c d e f g h")
            }
            print("$turn to move: ")
            val input = readLine()?.trim() ?: break
            if (input == "exit") { echo("Exiting game."); break }
            if (blind && input == "show") {
                for (r in 0 until 8) {
                    val row = board[r].joinToString(" ") { if (it.isEmpty()) "." else it }
                    echo("${8 - r} $row")
                }
                echo("  a b c d e f g h")
                continue
            }
            turn = if (turn == "White") "Black" else "White"
        }
    }
}

class ChessRandom : CliktCommand(name = "random", help = "Pick a random Chess960 starting position") {
    override fun run() {
        val positions = listOf(
            "BBQNNRKR", "BQNBNRKR", "BQNNRBKR", "BQNNRKRB", "QBBNNRKR", "QNBBNRKR", "QNBNRBKR", "QNBNRKRB",
            "QBNNBRKR", "QNNBBRKR", "QNNRBBKR", "QNNRBKRB", "QBNNRKBR", "QNNBRKBR", "QNNRKBBR", "QNNRKRBB"
        )
        val pos = positions[Random.nextInt(positions.size)]
        val idx = Random.nextInt(960) + 1
        val fen = "${pos.lowercase()}/pppppppp/8/8/8/8/PPPPPPPP/$pos w KQkq - 0 1"
        echo("Position $idx: $pos")
        echo("FEN: $fen")
    }
}

class ChessSetup : CliktCommand(name = "setup", help = "Set up a Chess960 starting position") {
    private val number by argument("position number").default("518")
    override fun run() {
        val positions = listOf(
            "BBQNNRKR", "BQNBNRKR", "BQNNRBKR", "BQNNRKRB", "QBBNNRKR", "QNBBNRKR", "QNBNRBKR", "QNBNRKRB",
            "QBNNBRKR", "QNNBBRKR", "QNNRBBKR", "QNNRBKRB", "QBNNRKBR", "QNNBRKBR", "QNNRKBBR", "QNNRKRBB"
        ) + (1..944).map { "BBQNNRKR" }
        val n = number.toInt()
        if (n < 1 || n > 960) throw Exception("position must be between 1 and 960")
        val idx = if (n - 1 < positions.size) n - 1 else 0
        val pos = positions[idx]
        val fen = "${pos.lowercase()}/pppppppp/8/8/8/8/PPPPPPPP/$pos w KQkq - 0 1"
        echo("Position $n: $pos")
        echo("FEN: $fen")
    }
}
